File: java/ContactBookJava/src/main/java/com/github/insanusmokrassar/KotlinVSJava/JavaComparisonModule/ContactBookJava/ContactBook.java
package com.github.insanusmokrassar.KotlinVSJava.JavaComparisonModule.ContactBookJava;

import com.github.insanusmokrassar.KotlinVSJava.JavaComparisonModule.ContactBookJava.API.Contact;
import com.github.insanusmokrassar.KotlinVSJava.JavaComparisonModule.ContactBookJava.API.ContactsBook;
import com.github.insanusmokrassar.KotlinVSJava.report_formatter.LogPoint;

import java.util.ArrayList;
import java.util.List;

import static com.github.insanusmokrassar.KotlinVSJava.report_formatter.LogPointKt.*;
import static com.github.insanusmokrassar.KotlinVSJava.report_formatter.TestDataFormatterKt.printPoints;

public class ContactBook {
    private final static Integer defaultBooksCount = 10;

    private final static Integer defaultContactsPerBook = 5000;

    private final static String testName = "ContactsBook.Java";

    public static void main(String[] args) {
        Integer booksCount = defaultBooksCount;

        if (args.length > 0) {
            try {
                booksCount = Integer.valueOf(args[0]);
            } catch (NumberFormatException ignored) {}
        }

        Integer contactsPerBook = defaultContactsPerBook;

        if (args.length > 1) {
            try {
                contactsPerBook = Integer.valueOf(args[1]);
            } catch (NumberFormatException ignored) {}
        }

        List<LogPoint> points = new ArrayList<>();

        points.add(new LogPoint(testName, START_TEST));

        List<ContactsBook> books = new ArrayList<>();

        for (int i = 0; i < booksCount; i++) {
            ContactsBook book = new ContactsBook();

            for (int contactIndex = 0; contactIndex < contactsPerBook; contactIndex++) {
                final int index = (contactIndex * 2) - i;

                book.add(
                        new Contact(
                                "Name " + index,
                                "Last Name " + index,
                                (index + 100) + "",
                                index + "@main.hello"
                        )
                );
            }

            books.add(book);
        }

        points.add(new LogPoint(testName, CONTACTS_BOOKS_FILLED));

        if (!books.isEmpty()) {
            ContactsBook firstBook = books.get(0);

            for (int i = 0; i < firstBook.size(); i++) {
                Contact contact = firstBook.get(i);

                for (ContactsBook book : books) {

                    if (book != firstBook) {
                        if (book.indexOf(contact) > i) {
                            break;
                        }
                    }
                }
            }
        }

        points.add(new LogPoint(testName, CONTACTS_BOOKS_SEARCH_COMPLETED));
        points.add(new LogPoint(testName, COMPLETE_TEST));

        printPoints(points);
    }
}

File: KotlinComparisonModule/ContactBookKotlin/src/main/kotlin/com/github/insanusmokrassar/KotlinVSJava/KotlinComparisonModule/ContactBookKotlin/ContactBook.kt
package com.github.insanusmokrassar.KotlinVSJava.KotlinComparisonModule.ContactBookKotlin

import com.github.insanusmokrassar.KotlinVSJava.KotlinComparisonModule.ContactBookKotlin.API.Contact
import com.github.insanusmokrassar.KotlinVSJava.KotlinComparisonModule.ContactBookKotlin.API.ContactsBook
import com.github.insanusmokrassar.KotlinVSJava.report_formatter.*

private const val defaultBooksCount = 10

private const val defaultContactsPerBook = 5000

private const val testName = "ContactsBook.Kotlin"

fun main(args: Array<String>) {
    val booksCount = if (args.isNotEmpty()) {
        args[0].toIntOrNull() ?: defaultBooksCount
    } else {
        defaultBooksCount
    }
    val contactsPerBook = if (args.size > 1) {
        args[1].toIntOrNull() ?: defaultContactsPerBook
    } else {
        defaultContactsPerBook
    }

    val points = ArrayList<LogPoint>()

    points.add(LogPoint(testName, START_TEST))

    val books = ArrayList<ContactsBook>()

    for (bookIndex in (0 until booksCount)) {
        ContactsBook().apply {
            for (contactIndex in (0 until contactsPerBook)) {
                val index = (contactIndex * 2) - bookIndex

                add(
                    Contact(
                        "Name $index",
                        "Last Name $index",
                        "${index + 100}",
                        "$index@mail.hello"
                    )
                )
            }
        }.also {
            books.add(it)
        }
    }

    points.add(LogPoint(testName, CONTACTS_BOOKS_FILLED))
    for (mainBook in books) {
        for ((i, contact) in mainBook.withIndex()) {
            val searchIndex = i * 2
            for (book in books) {

                if (book != mainBook) {
                    if (book.indexOf(contact) > searchIndex) {
                        break
                    }
                }

            }
        }
    }
    points.add(LogPoint(testName, CONTACTS_BOOKS_SEARCH_COMPLETED))
    points.add(LogPoint(testName, COMPLETE_TEST))

    printPoints(points)
}
